packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,172 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import { createRule, getParserServices } from '../util';
+
+export default createRule({
+  defaultOptions: [],
+  meta: {
+    docs: {
+      description: 'Disallow type parameters that only appear once',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      sole: 'Type parameter {{name}} is used only once.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  name: 'no-unnecessary-type-parameters',
+  create(context) {
+    const parserServices = getParserServices(context);
+    let usage: Map<ts.Identifier, tsutils.VariableInfo> | undefined;
+
+    return {
+      [[
+        'ClassDeclaration[typeParameters]',
+        'ClassExpression[typeParameters]',
+        'ArrowFunctionExpression[typeParameters]',
+        'FunctionDeclaration[typeParameters]',
+        'FunctionExpression[typeParameters]',
+        'TSCallSignatureDeclaration[typeParameters]',
+        'TSConstructorType[typeParameters]',
+        'TSDeclareFunction[typeParameters]',
+        'TSFunctionType[typeParameters]',
+        'TSMethodSignature[typeParameters]',
+        'TSEmptyBodyFunctionExpression[typeParameters]',
+      ].join(', ')](esNode: TSESTree.FunctionLike): void {
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(esNode) as
+          | ts.SignatureDeclaration
+          | ts.ClassLikeDeclaration;
+        if (!tsNode.typeParameters) {
+          return;
+        }
```

Comment:
[Refactor] We know `tsNode.typeParameters` must exist. So it'd be preferable to use a type assertion:

```suggestion
        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(
          esNode,
        ) as NodeWithTypeParameters;
```

There's a `MakeRequired` type in `no-unnecessary-type-constraint` that can be reused here nicely. https://github.com/typescript-eslint/typescript-eslint/pull/8019/files#diff-c2f1849228b57a79b7df2579a21fdd07b28dd4e734b9bb58e03aa9bf9f1ac7e4 also has it being  moved to a utils file.

```ts
type NodeWithTypeParameters = MakeRequired<
  ts.SignatureDeclaration | ts.ClassLikeDeclaration,
  'typeParameters'
>;
```